src/geode/io/geosciences/vo_input.cpp

```diff
@@ -0,0 +1,191 @@
+    class VOInputImpl
+    {
+    public:
+        VOInputImpl( absl::string_view filename, geode::RegularGrid3D& grid )
+            : file_{ geode::to_string( filename ) },
+              file_folder_{ geode::filepath_without_filename( filename ) },
+              grid_( grid ),
+              builder_{ geode::RegularGridBuilder3D::create( grid ) }
+        {
+            OPENGEODE_EXCEPTION(
+                file_.good(), "Error while opening file: ", filename );
+        }
+
+        void read_file()
+        {
+            geode::detail::check_keyword( file_, "GOCAD Voxet" );
+            const auto header = geode::detail::read_header( file_ );
+            builder_->set_name( header.name );
+            geode::detail::read_CRS( file_ );
+            initialize_grid();
+            read_data_file();
+        }
+
+    private:
+        void initialize_grid()
+        {
+            auto line = geode::detail::goto_keyword( file_, "AXIS_O" );
+            auto origin = read_coord( line, 1 );
+            const auto grid_size = read_grid_size( origin );
+            auto cells_number = read_cells_number();
+            auto cells_length = compute_cells_length( grid_size, cells_number );
+
+            builder_->initialize_grid(
+                origin, std::move( cells_number ), std::move( cells_length ) );
+        }
+
+        std::array< double, 3 > read_grid_size( const geode::Point3D& origin )
+        {
+            std::array< double, 3 > cells_length;
+            auto line = geode::detail::goto_keyword( file_, "AXIS_U" );
+            cells_length[0] =
+                geode::point_point_distance( origin, read_coord( line, 1 ) );
+            line = geode::detail::goto_keyword( file_, "AXIS_V" );
+            cells_length[1] =
+                geode::point_point_distance( origin, read_coord( line, 1 ) );
+            line = geode::detail::goto_keyword( file_, "AXIS_W" );
+            cells_length[2] =
+                geode::point_point_distance( origin, read_coord( line, 1 ) );
+            return cells_length;
+        }
+
+        std::array< geode::index_t, 3 > read_cells_number()
+        {
+            auto line = geode::detail::goto_keyword( file_, "AXIS_N" );
+            const auto tokens = geode::string_split( line );
+            return { geode::string_to_index( tokens[1] ),
+                geode::string_to_index( tokens[2] ),
+                geode::string_to_index( tokens[3] ) };
+        }
+
+        std::array< double, 3 > compute_cells_length(
+            const std::array< double, 3 >& grid_size,
+            const std::array< geode::index_t, 3 >& nb_cells )
+        {
+            std::array< double, 3 > result;
+            for( const auto axis_id : geode::LRange{ 3 } )
+            {
+                result[axis_id] = grid_size[axis_id] / nb_cells[axis_id];
+            }
+            return result;
+        }
+
+        geode::Point3D read_coord(
+            absl::string_view line, geode::index_t offset ) const
```

Comment:
This is not already done in this project?

Reply:
It is, I can homogenize it by adding ti to the utils.cpp